[Op] Add `StringTensorPack-15` and `StringTensorUnpack-15` specifications (#25104)
openvinotoolkit/openvino: docs/articles_en/documentation/openvino-ir-format/operation-sets/operation-specs/movement/string-tensor-pack-15.rst

.. {#openvino_docs_ops_type_StringTensorPack_15}

StringTensorPack
===================

.. meta::
  :description: Learn about StringTensorPack-15 - data movement operation which packs a concatenated batch of strings into a batched string tensor.

**Versioned name**: *StringTensorPack-15*

**Category**: *Data movement*

**Short description**: *StringTensorPack* operation uses outputs generated by *StringTensorUnpack* to transform a concatenated batch of strings into 
a string tensor.

**Detailed description**

Consider inputs:

* *word_begins* = [0, 5]
* *word_ends* = [5, 13]
* *output_symbols* = "IntelOpenVINO"

*StringTensorPack* uses indices from ``word_begins`` and ``word_ends`` to transform concatenated string ``output_symbols`` into ``output``, 
a batched string tensor. The ``output.shape`` (also called ``batch_size``) is equal to ``word_beings.shape`` and ``word_ends.shape``, 
and in this case holds values ``["Intel", "OpenVINO"]``.

**Inputs**

> * **1**: *word_begins*:

Reviewer: I think the `word` should be replaced by `string` (`string_begins`, `string_ends`) or dropped entirely (just `begins` and `ends`).

Author: Done.

  * **Description**: Indices of each string's begginings. **Required.**
  * **Shape**: 1D tensor of shape ``(batch_size)``.
  * **Type**: *T_IDX*

* **2**: *word_ends*:

  * **Description**: Indices of each string's endings. **Required.**
  * **Shape**: 1D tensor of shape ``(batch_size)``.
  * **Type**: *T_IDX*

> * **3**: *output_symbols*:

Reviewer: The word `output` in the input name might be confusing, so `symbols` or `chars` is better.

Author: Done.

**Types**

* *T*: ``string``.
* *T_IDX*: ``int64``.

**Examples**

*Example 1: input data as string*

.. code-block:: xml
   :force:

    <layer ... type="StringTensorPack" ... >
        <input>
            <port id="0" precision="I64">
                <dim>2</dim>     <!-- word_begins = [0, 5] -->
            </port>
            <port id="1" precision="I64">
                <dim>2</dim>     <!-- word_ends = [5, 13] -->
            </port>
            <port id="2" precision="string">
                <dim>13</dim>    <!-- output_symbols = "IntelOpenVINO" -->
            </port>
        </input>
        <output>
            <port id="0" precision="string">
                <dim>2</dim>     <!-- output = ["Intel", "OpenVINO"] -->
            </port>
        </output>
    </layer>

*Example 2: input with an empty string*

.. code-block:: xml
   :force:

    <layer ... type="StringTensorPack" ... >
        <input>
            <port id="0" precision="I64">
                <dim>2</dim>     <!-- word_begins = [0, 3, 3, 8, 9] -->
            </port>
            <port id="1" precision="I64">
                <dim>2</dim>     <!-- word_ends = [3, 3, 8, 9, 13] -->
            </port>

>             <port id="2" precision="string">

Reviewer: It is not a string tensor, it is `u8` tensor that holds utf-8 byte representation of concatenated strings.

Author: Fixed.

StringTensorPack
===================

.. meta::
  :description: Learn about StringTensorPack-15 - data movement operation which packs a concatenated batch of strings into a batched string tensor.

**Versioned name**: *StringTensorPack-15*

**Category**: *Data movement*

**Short description**: *StringTensorPack* transforms a concatenated batch of strings into 
a string tensor using *begins* and *ends* indices.

**Detailed description**

Consider inputs:

* *begins* = [0, 5]
* *ends* = [5, 13]
* *symbols* = "IntelOpenVINO"

*StringTensorPack* uses indices from ``begins`` and ``ends`` to transform concatenated string ``symbols`` into ``output``, 
a batched string tensor. The ``output.shape`` (also called ``batch_size``) is equal to ``begins.shape`` and ``ends.shape``, 
and in this case holds values ``["Intel", "OpenVINO"]``.

**Inputs**

* **1**: *begins*:

  * **Description**: Indices of each string's begginings. **Required.**
  * **Shape**: 1D tensor of shape ``(batch_size)``.
  * **Type**: *T_IDX*

* **2**: *ends*:

  * **Description**: Indices of each string's endings. **Required.**
  * **Shape**: 1D tensor of shape ``(batch_size)``.
  * **Type**: *T_IDX*

* **3**: *symbols*:

>   * **Description**: Concatenated ``input`` strings. **Required.**

Reviewer: ```suggestion
  * **Description**: Concatenated ``input`` strings encoded in utf-8 bytes. **Required.**
```

   :force:

    <layer ... type="StringTensorPack" ... >
        <input>
            <port id="0" precision="I64">
                <dim>2</dim>     <!-- begins = [0, 3, 3, 8, 9] -->
            </port>
            <port id="1" precision="I64">
                <dim>2</dim>     <!-- ends = [3, 3, 8, 9, 13] -->
            </port>
            <port id="2" precision="u8">
                <dim>13</dim>    <!-- symbols = "OMZGenAI 2024"-->
            </port>
        </input>
        <output>
            <port id="0" precision="u8">
                <dim>5</dim>     <!-- output = ["OMZ", "", "GenAI", " ", "2024"] -->
            </port>
        </output>
    </layer>

*Example 3: skipped symbols*

.. code-block:: xml
   :force:

    <layer ... type="StringTensorPack" ... >
        <input>
            <port id="0" precision="I64">
                <dim>2</dim>     <!-- begins = [0, 8] -->
            </port>
            <port id="1" precision="I64">
                <dim>2</dim>     <!-- ends = [1, 9] -->
            </port>
            <port id="2" precision="u8">
                <dim>13</dim>    <!-- symbols = "123456789"-->
            </port>
        </input>
        <output>

>             <port id="0" precision="u8">

Reviewer: Output is a string tensor, the `u8` precision is only int the input `symbols` tensor.

.. {#openvino_docs_ops_type_StringTensorPack_15}

StringTensorPack
===================

.. meta::
  :description: Learn about StringTensorPack-15 - data movement operation which packs a concatenated batch of strings into a batched string tensor.

**Versioned name**: *StringTensorPack-15*

**Category**: *Data movement*

**Short description**: *StringTensorPack* transforms a concatenated batch of strings into 
a string tensor using *begins* and *ends* indices.

**Detailed description**

Consider inputs:

* *begins* = [0, 5]
* *ends* = [5, 13]
* *symbols* = "IntelOpenVINO"

*StringTensorPack* uses indices from ``begins`` and ``ends`` to transform concatenated string ``symbols`` into ``output``, 
a batched string tensor. The ``output.shape`` (also called ``batch_size``) is equal to ``begins.shape`` and ``ends.shape``, 
and in this case holds values ``["Intel", "OpenVINO"]``.

When defining *begins* and *ends*, the notation ``[a, b)`` is used. This means that the range starts with ``a`` and includes all values up to, 
but not including, ``b``. That is why in the example given the length of "IntelOpenVINO" is 12, but *ends* vector contains 13.

**Inputs**

> * **1**: *begins*:

Reviewer 2: begins and ends are not 1D as it is said in `shape`. It can be ND tensor of the same shape as packed string tensor

Author: Corrected.

  * **Description**: Indices of each string's begginings. **Required.**
  * **Shape**: 1D tensor of shape ``(batch_size)``.
  * **Type**: *T_IDX*

* **2**: *ends*:

  * **Description**: Indices of each string's endings. **Required.**
  * **Shape**: 1D tensor of shape ``(batch_size)``.
  * **Type**: *T_IDX*

* **3**: *symbols*:

  * **Description**: Concatenated ``input`` strings encoded in utf-8 bytes. **Required.**
  * **Shape**: 1D tensor of shape equal to the total length of concatenated strings.
  * **Type**: *T*

**Outputs**

* **1**: *output*

  * **Description**: A string tensor.

>   * **Shape**: ``(batch_size)``

Reviewer 2: shape the same as `begins` and `ends`. It can be ND tensor

Author: Corrected.

>   * **Type**: *T*

Author: Fixed.

> .. {#openvino_docs_ops_type_StringTensorPack_15}

Reviewer 3: Remove file after moving to type category

StringTensorPack
===================

.. meta::
  :description: Learn about StringTensorPack-15 - data movement operation which packs a concatenated batch of strings into a batched string tensor.

**Versioned name**: *StringTensorPack-15*

> **Category**: *Data movement*

Author: Moved to "Type".

> **Short description**: *StringTensorPack* transforms a concatenated batch of strings into 
> a string tensor using *begins* and *ends* indices.

Reviewer 4: Can it be emphasized more that the input tensor is encoded? 
```suggestion
**Short description**: *StringTensorPack* transforms a concatenated strings data (encoded as 1D tensor of u8 element type) into 
a string tensor using *begins* and *ends* indices.
```

Consider inputs:

* *begins* = [0, 5]
* *ends* = [5, 13]
* *symbols* = "IntelOpenVINO"

*StringTensorPack* uses indices from ``begins`` and ``ends`` to transform concatenated string ``symbols`` into ``output``, 
a batched string tensor. The ``output.shape`` is equal to ``begins.shape`` and ``ends.shape``, 
and in this case ``output`` holds values ``["Intel", "OpenVINO"]``.

When defining *begins* and *ends*, the notation ``[a, b)`` is used. This means that the range starts with ``a`` and includes all values up to, 
but not including, ``b``. That is why in the example given the length of "IntelOpenVINO" is 12, but *ends* vector contains 13.

**Inputs**

* **1**: *begins*:

  * **Description**: Indices of each string's beginnings. **Required.**
  * **Range of values**: ND tensor of non-negative integer numbers.
  * **Type**: *T_IDX*

* **2**: *ends*:

  * **Description**: Indices of each string's endings. **Required.**
  * **Range of values**: ND tensor of non-negative integer numbers.
  * **Type**: *T_IDX*

* **3**: *symbols*:

  * **Description**: Concatenated ``input`` strings encoded in utf-8 bytes. **Required.**
  * **Range of values**: 1D tensor of element::string objects.

>   * **Type**: ``u8``

Reviewer 4: As already mentioned for unpack, the values are not "element::string objects".
Let's correct and align the description for both ops.

Author: Fixed for both.

**Outputs**

* **1**: *output*

  * **Description**: A string tensor.
  * **Range of values**: A string tensor.

>   * **Type**: ``element::string`

Reviewer 2: ```suggestion
  * **Type**: ``string`
```
